File: rdna3_kawpow/progpow.py
"""ProgPoW random-program generator (port of libprogpow/ProgPow.cpp).

For a program seed this builds the per-period random sequence ("the program") as
a backend-neutral intermediate representation (IR): a list of operations. The IR
is the single source of truth -- it is both:
  - rendered to GLSL (the GPU compute shader) / CUDA / OpenCL, and
  - executed directly in Python (the CPU correctness reference),
so the GPU and the reference are guaranteed to run the same program.

The exact ordering of kiss99() draws is what makes the program deterministic, so
the draw order here mirrors ProgPow.cpp precisely. Do not reorder.

For a given seed the rendered CUDA matches test/kernel.cu byte-for-byte (modulo
CUDA-equivalent cosmetics), which is the generator correctness check.
"""

from .constants import (
    PROGPOW_REGS, PROGPOW_DAG_LOADS, PROGPOW_CNT_CACHE, PROGPOW_CNT_MATH,
    PROGPOW_CACHE_WORDS, FNV_PRIME, FNV_OFFSET_BASIS, MASK32,
)

CUDA = "cuda"
CL = "cl"
GLSL = "glsl"


class _Kiss99:
    """KISS99 PRNG, matching ProgPow.cpp (32-bit wraparound)."""

    __slots__ = ("z", "w", "jsr", "jcong")

    def __init__(self, z, w, jsr, jcong):
        self.z, self.w, self.jsr, self.jcong = z, w, jsr, jcong

    def __call__(self):
        self.z = (36969 * (self.z & 0xFFFF) + (self.z >> 16)) & MASK32
        self.w = (18000 * (self.w & 0xFFFF) + (self.w >> 16)) & MASK32
        mwc = ((self.z << 16) + self.w) & MASK32
        self.jsr ^= (self.jsr << 17) & MASK32
        self.jsr ^= (self.jsr >> 13)
        self.jsr ^= (self.jsr << 5) & MASK32
        self.jsr &= MASK32
        self.jcong = (69069 * self.jcong + 1234567) & MASK32
        return ((mwc ^ self.jcong) + self.jsr) & MASK32


def _fnv1a(h, d):
    return ((h ^ d) * FNV_PRIME) & MASK32


def build_program(prog_seed, cnt_cache=PROGPOW_CNT_CACHE, cnt_math=PROGPOW_CNT_MATH):
    """Build the IR for `prog_seed`.

    Returns a list of op tuples:
      ("global",)                         global DAG load (shuffle + index)
      ("cache", src, dst, r)              cache load then merge(mix[dst], data, r)
      ("math",  s1, s2, dst, r1, r2)      data=math(mix[s1],mix[s2],r1); merge(dst,r2)
      ("dagmerge", dst, i, r)             merge(mix[dst], data_dag.s[i], r)

    cnt_cache/cnt_math default to production KawPow (11/18); pass 12/20 to
    reproduce the ProgPoW-0.9.2 reference in test/kernel.cu.
    """
    seed0 = prog_seed & MASK32
    seed1 = (prog_seed >> 32) & MASK32
    fnv_hash = FNV_OFFSET_BASIS
    fnv_hash = _fnv1a(fnv_hash, seed0); z = fnv_hash
    fnv_hash = _fnv1a(fnv_hash, seed1); w = fnv_hash
    fnv_hash = _fnv1a(fnv_hash, seed0); jsr = fnv_hash
    fnv_hash = _fnv1a(fnv_hash, seed1); jcong = fnv_hash
    rnd = _Kiss99(z, w, jsr, jcong)

    mix_seq_dst = list(range(PROGPOW_REGS))
    mix_seq_cache = list(range(PROGPOW_REGS))
    for i in range(PROGPOW_REGS - 1, 0, -1):
        j = rnd() % (i + 1)
        mix_seq_dst[i], mix_seq_dst[j] = mix_seq_dst[j], mix_seq_dst[i]
        j = rnd() % (i + 1)
        mix_seq_cache[i], mix_seq_cache[j] = mix_seq_cache[j], mix_seq_cache[i]

    dst_cnt = cache_cnt = 0
    ops = [("global",)]

    for i in range(max(cnt_cache, cnt_math)):
        if i < cnt_cache:
            src = mix_seq_cache[cache_cnt % PROGPOW_REGS]; cache_cnt += 1
            dst = mix_seq_dst[dst_cnt % PROGPOW_REGS]; dst_cnt += 1
            r = rnd()
            ops.append(("cache", src, dst, r))
        if i < cnt_math:
            src_rnd = rnd() % ((PROGPOW_REGS - 1) * PROGPOW_REGS)
            src1 = src_rnd % PROGPOW_REGS
            src2 = src_rnd // PROGPOW_REGS
            if src2 >= src1:
                src2 += 1
            r1 = rnd()
            dst = mix_seq_dst[dst_cnt % PROGPOW_REGS]; dst_cnt += 1
            r2 = rnd()
            ops.append(("math", src1, src2, dst, r1, r2))

    ops.append(("dagmerge", 0, 0, rnd()))
    for i in range(1, PROGPOW_DAG_LOADS):
        dst = mix_seq_dst[dst_cnt % PROGPOW_REGS]; dst_cnt += 1
        ops.append(("dagmerge", dst, i, rnd()))
    return ops


# --- math / merge op selection (shared by renderer and interpreter) --------------

def _math_sel(r):
    return r % 11


def _merge_sel(r):
    return r % 4


def _merge_rot(r):
    return ((r >> 16) % 31) + 1


# ================================ TEXT RENDERER ==================================

def _math_text(d, a, b, r):
    sel = _math_sel(r)
    return [
        f"{d} = {a} + {b};",
        f"{d} = {a} * {b};",
        f"{d} = mul_hi({a}, {b});",
        f"{d} = min({a}, {b});",
        f"{d} = ROTL32({a}, {b} % 32);",
        f"{d} = ROTR32({a}, {b} % 32);",
        f"{d} = {a} & {b};",
        f"{d} = {a} | {b};",
        f"{d} = {a} ^ {b};",
        f"{d} = clz({a}) + clz({b});",
        f"{d} = popcount({a}) + popcount({b});",
    ][sel]


def _merge_text(a, b, r):
    sel = _merge_sel(r)
    if sel == 0:
        return f"{a} = ({a} * 33) + {b};"
    if sel == 1:
        return f"{a} = ({a} ^ {b}) * 33;"
    if sel == 2:
        return f"{a} = ROTL32({a}, {_merge_rot(r)}) ^ {b};"
    return f"{a} = ROTR32({a}, {_merge_rot(r)}) ^ {b};"


def render_loop(prog_seed, kern, ops=None, cnt_cache=PROGPOW_CNT_CACHE,
                cnt_math=PROGPOW_CNT_MATH):
    """Render the progPowLoop source for backend `kern` from IR `ops`."""
    if ops is None:
        ops = build_program(prog_seed, cnt_cache, cnt_math)
    out = [_signature(prog_seed, kern), "{\n", "dag_t data_dag;\n",
           "uint32_t offset, data;\n"]
    if kern == CL:
        out.append("uint32_t mix[PROGPOW_REGS];\n"
                   "for(int i=0; i<PROGPOW_REGS; i++)\n    mix[i] = mix_arg[i];\n")
    out.append(_lane_id(kern))

    cache_i = math_i = 0
    for op in ops:
        if op[0] == "global":
            out.append("// global load\n")
            out.append(_global_shuffle(kern))
            out.append("offset %= PROGPOW_DAG_ELEMENTS;\n")
            out.append("offset = offset * PROGPOW_LANES + (lane_id ^ loop) % PROGPOW_LANES;\n")
            if kern == GLSL:
                # g_dag is a flat uint[] SSBO: dag_t row r occupies words [r*4 .. r*4+3].
                for i in range(PROGPOW_DAG_LOADS):
                    out.append(f"data_dag.s[{i}] = g_dag[offset * PROGPOW_DAG_LOADS + {i}u];\n")
            else:
                out.append("data_dag = g_dag[offset];\n")
            out.append("// hack to prevent compiler from reordering LD and usage\n")
            out.append(_fence(kern))
        elif op[0] == "cache":
            _, src, dst, r = op
            out.append(f"// cache load {cache_i}\n"); cache_i += 1
            out.append(f"offset = mix[{src}] % PROGPOW_CACHE_WORDS;\n")
            out.append("data = c_dag[offset];\n")
            out.append(_merge_text(f"mix[{dst}]", "data", r) + "\n")
        elif op[0] == "math":
            _, s1, s2, dst, r1, r2 = op
            out.append(f"// random math {math_i}\n"); math_i += 1
            out.append(_math_text("data", f"mix[{s1}]", f"mix[{s2}]", r1) + "\n")
            out.append(_merge_text(f"mix[{dst}]", "data", r2) + "\n")
        elif op[0] == "dagmerge":
            _, dst, i, r = op
            if i == 0:
                out.append("// consume global load data\n")
                out.append("// hack to prevent compiler from reordering LD and usage\n")
                out.append(_fence(kern))
            out.append(_merge_text(f"mix[{dst}]", f"data_dag.s[{i}]", r) + "\n")

    if kern == CL:
        out.append("for(int i=0; i<PROGPOW_REGS; i++)\n    mix_arg[i] = mix[i];\n")
    out.append("}\n")
    return "".join(out)


# Backwards-compatible alias used by earlier validation
def generate_loop(prog_seed, kern, cnt_cache=PROGPOW_CNT_CACHE, cnt_math=PROGPOW_CNT_MATH):
    return render_loop(prog_seed, kern, None, cnt_cache, cnt_math)


def _signature(prog_seed, kern):
    if kern == CUDA:
        return (f"// Inner loop for prog_seed {prog_seed}\n"
                "__device__ __forceinline__ void progPowLoop(const uint32_t loop,\n"
                "    uint32_t mix[PROGPOW_REGS],\n"
                "    const dag_t *g_dag,\n"
                "    const uint32_t c_dag[PROGPOW_CACHE_WORDS],\n"
                "    const bool hack_false)\n")
    if kern == CL:
        return (f"// Inner loop for prog_seed {prog_seed}\n"
                "inline void progPowLoop(const uint32_t loop,\n"
                "        volatile uint32_t mix_arg[PROGPOW_REGS],\n"
                "        __global const dag_t *g_dag,\n"
                "        __local const uint32_t c_dag[PROGPOW_CACHE_WORDS],\n"
                "        __local uint64_t share[GROUP_SHARE],\n"
                "        const bool hack_false)\n")
    return (f"// Inner loop for prog_seed {prog_seed}\n"
            "void progPowLoop(const uint32_t loop, inout uint32_t mix[PROGPOW_REGS],\n"
            "        uint32_t lane_id, uint32_t sg_base, bool hack_false)\n")


def _lane_id(kern):
    if kern == CUDA:
        return "const uint32_t lane_id = threadIdx.x & (PROGPOW_LANES-1);\n"
    if kern == CL:
        return ("const uint32_t lane_id = get_local_id(0) & (PROGPOW_LANES-1);\n"
                "const uint32_t group_id = get_local_id(0) / PROGPOW_LANES;\n")
    return ""  # GLSL: lane_id / sg_base passed in by caller


def _global_shuffle(kern):
    if kern == CUDA:
        return "offset = SHFL(mix[0], loop%PROGPOW_LANES, PROGPOW_LANES);\n"
    if kern == CL:
        return ("if(lane_id == (loop % PROGPOW_LANES))\n    share[group_id] = mix[0];\n"
                "barrier(CLK_LOCAL_MEM_FENCE);\noffset = share[group_id];\n")
    # GLSL/RDNA3: subgroupShuffle straight from registers -- no LDS, no barrier.
    return "offset = subgroupShuffle(mix[0], sg_base + (loop % PROGPOW_LANES));\n"


def _fence(kern):
    if kern == CUDA:
        return "if (hack_false) __threadfence_block();\n"
    if kern == CL:
        return "if (hack_false) barrier(CLK_LOCAL_MEM_FENCE);\n"
    return "if (hack_false) subgroupBarrier();\n"


# ================================ PYTHON INTERPRETER =============================

def _rotl32(x, n):
    n &= 31
    return ((x << n) | (x >> (32 - n))) & MASK32 if n else x & MASK32


def _rotr32(x, n):
    n &= 31
    return ((x >> n) | (x << (32 - n))) & MASK32 if n else x & MASK32


def _mul_hi(a, b):
    return ((a * b) >> 32) & MASK32


def _clz(x):
    return 32 if x == 0 else 31 - x.bit_length() + 1  # 32 - bit_length


def _popcount(x):
    return bin(x & MASK32).count("1")


def _math_exec(a, b, r):
    sel = _math_sel(r)
    a &= MASK32; b &= MASK32
    if sel == 0:
        return (a + b) & MASK32
    if sel == 1:
        return (a * b) & MASK32
    if sel == 2:
        return _mul_hi(a, b)
    if sel == 3:
        return min(a, b)
    if sel == 4:
        return _rotl32(a, b % 32)
    if sel == 5:
        return _rotr32(a, b % 32)
    if sel == 6:
        return a & b
    if sel == 7:
        return a | b
    if sel == 8:
        return a ^ b
    if sel == 9:
        return (_clz(a) + _clz(b)) & MASK32
    return (_popcount(a) + _popcount(b)) & MASK32


def _merge_exec(a, b, r):
    sel = _merge_sel(r)
    a &= MASK32; b &= MASK32
    if sel == 0:
        return ((a * 33) + b) & MASK32
    if sel == 1:
        return ((a ^ b) * 33) & MASK32
    if sel == 2:
        return _rotl32(a, _merge_rot(r)) ^ b
    return _rotr32(a, _merge_rot(r)) ^ b


def run_loop(ops, loop, mix_lanes, c_dag, dag_entry_for_loop, dag_elements):
    """Execute one progPowLoop iteration across all 16 lanes (CPU reference).

    mix_lanes: list[16] of list[32] uint32 (modified in place).
    c_dag: the 4096-word cache.
    dag_entry_for_loop(entry_index) -> list[PROGPOW_DAG_LOADS] uint32 (one DAG row).
    """
    from .constants import PROGPOW_LANES, PROGPOW_DAG_LOADS

    # Global load: lane (loop % LANES) broadcasts mix[0]; each lane reads its DAG row.
    src_lane = loop % PROGPOW_LANES
    base_offset = mix_lanes[src_lane][0] % dag_elements
    data_dag = []
    for lane in range(PROGPOW_LANES):
        off = base_offset * PROGPOW_LANES + ((lane ^ loop) % PROGPOW_LANES)
        data_dag.append(dag_entry_for_loop(off))

    for lane in range(PROGPOW_LANES):
        mix = mix_lanes[lane]
        dd = data_dag[lane]
        for op in ops:
            if op[0] == "global":
                continue
            if op[0] == "cache":
                _, src, dst, r = op
                data = c_dag[mix[src] % PROGPOW_CACHE_WORDS]
                mix[dst] = _merge_exec(mix[dst], data, r)
            elif op[0] == "math":
                _, s1, s2, dst, r1, r2 = op
                data = _math_exec(mix[s1], mix[s2], r1)
                mix[dst] = _merge_exec(mix[dst], data, r2)
            elif op[0] == "dagmerge":
                _, dst, i, r = op
                mix[dst] = _merge_exec(mix[dst], dd[i], r)

File: rust/src/progpow.rs
//! ProgPoW random-program generator (port of `rdna3_kawpow/progpow.py`).
//!
//! For a program seed this builds the per-period random sequence ("the program")
//! as a backend-neutral IR (`Vec<Op>`). That IR is the single source of truth: it
//! is both rendered to GLSL (the GPU shader) and executed directly here (the CPU
//! reference), so the GPU and the reference run the identical program.
//!
//! The exact ordering of `Kiss99` draws is what makes the program deterministic,
//! so the draw order here mirrors `ProgPow.cpp` precisely. DO NOT REORDER DRAWS.

use std::fmt::Write as _;

use crate::constants::{
    FNV_OFFSET_BASIS, FNV_PRIME, PROGPOW_CNT_CACHE, PROGPOW_CNT_MATH, PROGPOW_DAG_LOADS,
    PROGPOW_REGS,
};

/// One operation in the per-period program IR.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Op {
    /// Global DAG load (shuffle + index).
    Global,
    /// Cache load then merge(mix[dst], data, r).
    Cache { src: usize, dst: usize, r: u32 },
    /// data = math(mix[s1], mix[s2], r1); merge(mix[dst], data, r2).
    Math {
        s1: usize,
        s2: usize,
        dst: usize,
        r1: u32,
        r2: u32,
    },
    /// merge(mix[dst], data_dag.s[i], r).
    DagMerge { dst: usize, i: usize, r: u32 },
}

/// KISS99 PRNG, matching `ProgPow.cpp` (32-bit wraparound).
#[derive(Clone, Copy)]
pub struct Kiss99 {
    z: u32,
    w: u32,
    jsr: u32,
    jcong: u32,
}

impl Kiss99 {
    pub fn new(z: u32, w: u32, jsr: u32, jcong: u32) -> Self {
        Kiss99 { z, w, jsr, jcong }
    }

    pub fn next_u32(&mut self) -> u32 {
        self.z = (36969u32.wrapping_mul(self.z & 0xFFFF)).wrapping_add(self.z >> 16);
        self.w = (18000u32.wrapping_mul(self.w & 0xFFFF)).wrapping_add(self.w >> 16);
        let mwc = (self.z << 16).wrapping_add(self.w);
        self.jsr ^= self.jsr << 17;
        self.jsr ^= self.jsr >> 13;
        self.jsr ^= self.jsr << 5;
        self.jcong = 69069u32.wrapping_mul(self.jcong).wrapping_add(1234567);
        (mwc ^ self.jcong).wrapping_add(self.jsr)
    }
}

#[inline]
fn fnv1a(h: u32, d: u32) -> u32 {
    (h ^ d).wrapping_mul(FNV_PRIME)
}

/// Build the IR for `prog_seed`. `cnt_cache`/`cnt_math` default to production
/// KawPow (11/18); pass 12/20 to reproduce the ProgPoW-0.9.2 reference.
pub fn build_program(prog_seed: u64, cnt_cache: usize, cnt_math: usize) -> Vec<Op> {
    let seed0 = prog_seed as u32;
    let seed1 = (prog_seed >> 32) as u32;
    let mut fnv_hash = FNV_OFFSET_BASIS;
    fnv_hash = fnv1a(fnv_hash, seed0);
    let z = fnv_hash;
    fnv_hash = fnv1a(fnv_hash, seed1);
    let w = fnv_hash;
    fnv_hash = fnv1a(fnv_hash, seed0);
    let jsr = fnv_hash;
    fnv_hash = fnv1a(fnv_hash, seed1);
    let jcong = fnv_hash;
    let mut rnd = Kiss99::new(z, w, jsr, jcong);

    let mut mix_seq_dst: Vec<usize> = (0..PROGPOW_REGS).collect();
    let mut mix_seq_cache: Vec<usize> = (0..PROGPOW_REGS).collect();
    for i in (1..PROGPOW_REGS).rev() {
        let j = (rnd.next_u32() as usize) % (i + 1);
        mix_seq_dst.swap(i, j);
        let j = (rnd.next_u32() as usize) % (i + 1);
        mix_seq_cache.swap(i, j);
    }

    let mut dst_cnt = 0usize;
    let mut cache_cnt = 0usize;
    let mut ops = vec![Op::Global];

    for i in 0..cnt_cache.max(cnt_math) {
        if i < cnt_cache {
            let src = mix_seq_cache[cache_cnt % PROGPOW_REGS];
            cache_cnt += 1;
            let dst = mix_seq_dst[dst_cnt % PROGPOW_REGS];
            dst_cnt += 1;
            let r = rnd.next_u32();
            ops.push(Op::Cache { src, dst, r });
        }
        if i < cnt_math {
            let regs = PROGPOW_REGS as u32;
            let src_rnd = rnd.next_u32() % ((regs - 1) * regs);
            let src1 = (src_rnd % regs) as usize;
            let mut src2 = (src_rnd / regs) as usize;
            if src2 >= src1 {
                src2 += 1;
            }
            let r1 = rnd.next_u32();
            let dst = mix_seq_dst[dst_cnt % PROGPOW_REGS];
            dst_cnt += 1;
            let r2 = rnd.next_u32();
            ops.push(Op::Math {
                s1: src1,
                s2: src2,
                dst,
                r1,
                r2,
            });
        }
    }

    ops.push(Op::DagMerge {
        dst: 0,
        i: 0,
        r: rnd.next_u32(),
    });
    for i in 1..PROGPOW_DAG_LOADS {
        let dst = mix_seq_dst[dst_cnt % PROGPOW_REGS];
        dst_cnt += 1;
        ops.push(Op::DagMerge {
            dst,
            i,
            r: rnd.next_u32(),
        });
    }
    ops
}

/// Build with production-default op counts.
pub fn build_program_default(prog_seed: u64) -> Vec<Op> {
    build_program(prog_seed, PROGPOW_CNT_CACHE, PROGPOW_CNT_MATH)
}

// --- math / merge op selection (shared by renderer and interpreter) ---

#[inline]
fn math_sel(r: u32) -> u32 {
    r % 11
}
#[inline]
fn merge_sel(r: u32) -> u32 {
    r % 4
}
#[inline]
fn merge_rot(r: u32) -> u32 {
    ((r >> 16) % 31) + 1
}

// ================================ GLSL RENDERER ==================================

fn math_text(d: &str, a: &str, b: &str, r: u32) -> String {
    match math_sel(r) {
        0 => format!("{d} = {a} + {b};"),
        1 => format!("{d} = {a} * {b};"),
        2 => format!("{d} = mul_hi({a}, {b});"),
        3 => format!("{d} = min({a}, {b});"),
        4 => format!("{d} = ROTL32({a}, {b} % 32);"),
        5 => format!("{d} = ROTR32({a}, {b} % 32);"),
        6 => format!("{d} = {a} & {b};"),
        7 => format!("{d} = {a} | {b};"),
        8 => format!("{d} = {a} ^ {b};"),
        9 => format!("{d} = clz({a}) + clz({b});"),
        _ => format!("{d} = popcount({a}) + popcount({b});"),
    }
}

fn merge_text(a: &str, b: &str, r: u32) -> String {
    match merge_sel(r) {
        0 => format!("{a} = ({a} * 33) + {b};"),
        1 => format!("{a} = ({a} ^ {b}) * 33;"),
        2 => format!("{a} = ROTL32({a}, {}) ^ {b};", merge_rot(r)),
        _ => format!("{a} = ROTR32({a}, {}) ^ {b};", merge_rot(r)),
    }
}

/// Render the `progPowLoop` GLSL source for `ops` (which must come from
/// `build_program(prog_seed, ...)`). Byte-identical to the Python GLSL renderer.
pub fn render_loop_glsl(prog_seed: u64, ops: &[Op]) -> String {
    let mut out = String::new();
    let _ = write!(
        out,
        "// Inner loop for prog_seed {prog_seed}\n\
         void progPowLoop(const uint32_t loop, inout uint32_t mix[PROGPOW_REGS],\n        \
         uint32_t lane_id, uint32_t sg_base, bool hack_false)\n"
    );
    out.push_str("{\n");
    out.push_str("dag_t data_dag;\n");
    out.push_str("uint32_t offset, data;\n");

    let mut cache_i = 0usize;
    let mut math_i = 0usize;
    for op in ops {
        match *op {
            Op::Global => {
                out.push_str("// global load\n");
                out.push_str(
                    "offset = subgroupShuffle(mix[0], sg_base + (loop % PROGPOW_LANES));\n",
                );
                out.push_str("offset %= PROGPOW_DAG_ELEMENTS;\n");
                out.push_str(
                    "offset = offset * PROGPOW_LANES + (lane_id ^ loop) % PROGPOW_LANES;\n",
                );
                for i in 0..PROGPOW_DAG_LOADS {
                    let _ = writeln!(
                        out,
                        "data_dag.s[{i}] = g_dag[offset * PROGPOW_DAG_LOADS + {i}u];"
                    );
                }
                out.push_str("// hack to prevent compiler from reordering LD and usage\n");
                out.push_str("if (hack_false) subgroupBarrier();\n");
            }
            Op::Cache { src, dst, r } => {
                let _ = writeln!(out, "// cache load {cache_i}");
                cache_i += 1;
                let _ = writeln!(out, "offset = mix[{src}] % PROGPOW_CACHE_WORDS;");
                out.push_str("data = c_dag[offset];\n");
                let _ = writeln!(out, "{}", merge_text(&format!("mix[{dst}]"), "data", r));
            }
            Op::Math { s1, s2, dst, r1, r2 } => {
                let _ = writeln!(out, "// random math {math_i}");
                math_i += 1;
                let _ = writeln!(
                    out,
                    "{}",
                    math_text("data", &format!("mix[{s1}]"), &format!("mix[{s2}]"), r1)
                );
                let _ = writeln!(out, "{}", merge_text(&format!("mix[{dst}]"), "data", r2));
            }
            Op::DagMerge { dst, i, r } => {
                if i == 0 {
                    out.push_str("// consume global load data\n");
                    out.push_str("// hack to prevent compiler from reordering LD and usage\n");
                    out.push_str("if (hack_false) subgroupBarrier();\n");
                }
                let _ = writeln!(
                    out,
                    "{}",
                    merge_text(&format!("mix[{dst}]"), &format!("data_dag.s[{i}]"), r)
                );
            }
        }
    }
    out.push_str("}\n");
    out
}

// ================================ INTERPRETER ===================================

#[inline]
fn mul_hi(a: u32, b: u32) -> u32 {
    (((a as u64) * (b as u64)) >> 32) as u32
}

fn math_exec(a: u32, b: u32, r: u32) -> u32 {
    match math_sel(r) {
        0 => a.wrapping_add(b),
        1 => a.wrapping_mul(b),
        2 => mul_hi(a, b),
        3 => a.min(b),
        4 => a.rotate_left(b % 32),
        5 => a.rotate_right(b % 32),
        6 => a & b,
        7 => a | b,
        8 => a ^ b,
        9 => a.leading_zeros().wrapping_add(b.leading_zeros()),
        _ => a.count_ones().wrapping_add(b.count_ones()),
    }
}

fn merge_exec(a: u32, b: u32, r: u32) -> u32 {
    match merge_sel(r) {
        0 => a.wrapping_mul(33).wrapping_add(b),
        1 => (a ^ b).wrapping_mul(33),
        2 => a.rotate_left(merge_rot(r)) ^ b,
        _ => a.rotate_right(merge_rot(r)) ^ b,
    }
}

/// Execute one `progPowLoop` iteration across all 16 lanes (CPU reference).
///
/// `mix_lanes`: 16 lanes x 32 u32 (modified in place). `cdag`: the 4096-word cache.
/// `dag_row(entry_index)` returns one 4-word DAG row.
pub fn run_loop<F: Fn(u32) -> [u32; 4]>(
    ops: &[Op],
    loop_idx: u32,
    mix_lanes: &mut [[u32; crate::constants::PROGPOW_REGS]],
    cdag: &[u32],
    dag_row: F,
    dag_elements: u32,
) {
    use crate::constants::{PROGPOW_CACHE_WORDS, PROGPOW_LANES};
    let lanes = PROGPOW_LANES;

    let src_lane = (loop_idx % lanes) as usize;
    let base_offset = (mix_lanes[src_lane][0] as u64 % dag_elements as u64) as u32;
    let mut data_dag = [[0u32; 4]; 16];
    for lane in 0..lanes {
        let off = base_offset as u64 * lanes as u64 + ((lane ^ loop_idx) % lanes) as u64;
        data_dag[lane as usize] = dag_row(off as u32);
    }

    let cache_words = PROGPOW_CACHE_WORDS as u32;
    for lane in 0..lanes as usize {
        let dd = data_dag[lane];
        let mix = &mut mix_lanes[lane];
        for op in ops {
            match *op {
                Op::Global => {}
                Op::Cache { src, dst, r } => {
                    let data = cdag[(mix[src] % cache_words) as usize];
                    mix[dst] = merge_exec(mix[dst], data, r);
                }
                Op::Math { s1, s2, dst, r1, r2 } => {
                    let data = math_exec(mix[s1], mix[s2], r1);
                    mix[dst] = merge_exec(mix[dst], data, r2);
                }
                Op::DagMerge { dst, i, r } => {
                    mix[dst] = merge_exec(mix[dst], dd[i], r);
                }
            }
        }
    }
}
